src/hotspot/share/opto/vectornode.cpp

```diff
@@ -1299,6 +1299,89 @@ Node* ReductionNode::Ideal(PhaseGVN* phase, bool can_reshape) {
   return nullptr;
 }
 
+Node* UnorderedReductionNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  Node* n = ReductionNode::Ideal(phase, can_reshape);
+  if (n != nullptr) {
+    return n;
+  }
+  if (can_reshape) {
+    // Having a ReductionNode in the loop is expensive. It needs to recursively
+    // fold together the vector values, for every vectorized loop iteration. If
+    // we encounter the following pattern, we can move the UnorderedReduction
+    // outside the loop.
+    //
+    // CountedLoop     init
+    //          |        |
+    //          +------+ | +---------------+
+    //                 | | |               |
+    //                PhiNode (s)  Vector  |
+    //                  |            |     |
+    //               UnorderedReduction    |
+    //                       |             |
+    //                       +-------------+
+    //
+    // We patch the graph to look like this:
+    //
+    // CountedLoop   neutral_vector
+    //         |         |
+    //         +-------+ | +---------------+
+    //                 | | |               |
+    //                PhiNode (v)  Vector  |
+    //                   |           |     |
+    //      init       VectorAccumulator   |
+    //        |          |     |           |
+    //     UnorderedReduction  +-----------+
+    //
+    // We turned the scalar (s) Phi into a vectorized one (v). In the loop, we
+    // use a vector_accumulator, which does the same reduction, but only element
+    // wise. This is a single operation, rather than many for the ReductionNode.
+    // We can then reduce that vector_accumulator after the loop, and also reduce
+    // the init value into it.
+    // We can not do this with all reductions. Some reductions do not allow the
+    // reordering of operations (for example float addition).
+    Node* ctrl = in(0);
+    Node* in1 = in(1);
+    Node* in2 = in(2);
+    if (ctrl == nullptr &&
+        in1 != nullptr && in1->is_Phi() && in1->in(2) == this && in1->outcnt() == 1 &&
+        in1->in(0)->is_CountedLoop() &&
+        in2->is_Vector()) {
```

Comment:
Should you also check that this reduction node doesn't have users inside loop?

Reply:
@vnkozlov How should I do that? Can that even be done during IGVN? Or should I move the implementation to loopopts?

Reply:
For example, I could put it before or after `try_sink_out_of_loop` inside `PhaseIdealLoop::split_if_with_blocks_post`.

Reply:
As we discussed offline, you may check and mark Reduction node if it has users inside loop in SuperWord code where we are creating vector referenced by Reduction node.